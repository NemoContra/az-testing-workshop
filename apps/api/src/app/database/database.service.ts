import { Injectable } from '@nestjs/common';
import { Contract } from '../models/contract.model';

@Injectable()
export class DatabaseService {
  private mockData: Contract[] = [
    {
      id: '123456789',
      contractNumber: '1/2345678/9',
      start: '2024-01-01',
      premium: 42.42,
      person: { firstname: 'Homer', lastname: 'Simpson', dateOfBirth: '1961-05-16' },
    },
    {
      id: '123456788',
      contractNumber: '1/2345678/8',
      start: '2024-02-01',
      premium: 99.000,
      person: { firstname: 'Bart', lastname: 'Simpson', dateOfBirth: '1995-08-21' },
    },
  ];

  public getAllContracts(): Promise<Contract[]> {
    return Promise.resolve(this.mockData);
  }

  public getContractById(contractId: string): Promise<Contract | undefined> {
    return Promise.resolve(this.mockData.find(({ id }) => id === contractId));
  }

  public updateContract(updatedContract: Partial<Contract>): Promise<Contract | undefined> {
    const existingContract = this.mockData.find(({ id }) => id === updatedContract.id);

    if (!existingContract) {
      return Promise.resolve(undefined);
    }

    const updatedContracts = this.mockData.filter(({ id }) => id !== updatedContract.id);

    this.mockData = [...updatedContracts, {...existingContract, ...updatedContract}];

    return Promise.resolve({...existingContract, ...updatedContract});
  }
}
