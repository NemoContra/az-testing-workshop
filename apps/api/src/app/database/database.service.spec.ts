import { Test, TestingModule } from '@nestjs/testing';
import { DatabaseService } from './database.service';

describe('DatabaseService', () => {
  let service: DatabaseService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [DatabaseService],
    }).compile();

    service = module.get<DatabaseService>(DatabaseService);
  });

  describe('getAllContracts', () => {
    it('should return all contracts', async () => {
      await expect(service.getAllContracts()).resolves.toMatchSnapshot();
    });
  });

  describe('getContractById', () => {
    it('should return contract by id', async () => {
      await expect(service.getContractById('123456789')).resolves.toMatchSnapshot();
    });

    it('should return undefined if id is not existing', async () => {
      await expect(service.getContractById('123')).resolves.toEqual(undefined);
    });
  });

  describe('updateContract', () => {
    it('should return contract by id', async () => {
      await expect(service.updateContract( {
        id: '123456788',
        person: { firstname: 'Bart', lastname: 'Stark', dateOfBirth: '1995-08-21' },
      })).resolves.toMatchSnapshot();
    });

    it('should return undefined if contract to update is not found', async () => {
      await expect(service.updateContract({
        id: '123',
        person: { firstname: 'Bart', lastname: 'Stark', dateOfBirth: '1995-08-21' },
      })).resolves.toEqual(undefined);
    });
  });

});
