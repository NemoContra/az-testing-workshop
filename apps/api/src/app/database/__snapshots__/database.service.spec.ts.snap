// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatabaseService getAllContracts should return all contracts 1`] = `
[
  {
    "contractNumber": "1/2345678/9",
    "id": "123456789",
    "person": {
      "dateOfBirth": "1961-05-16",
      "firstname": "Homer",
      "lastname": "Simpson",
    },
    "premium": 42.42,
    "start": "2024-01-01",
  },
  {
    "contractNumber": "1/2345678/8",
    "id": "123456788",
    "person": {
      "dateOfBirth": "1995-08-21",
      "firstname": "Bart",
      "lastname": "Simpson",
    },
    "premium": 99,
    "start": "2024-02-01",
  },
]
`;

exports[`DatabaseService getContractById should return contract by id 1`] = `
{
  "contractNumber": "1/2345678/9",
  "id": "123456789",
  "person": {
    "dateOfBirth": "1961-05-16",
    "firstname": "Homer",
    "lastname": "Simpson",
  },
  "premium": 42.42,
  "start": "2024-01-01",
}
`;

exports[`DatabaseService updateContract should return contract by id 1`] = `
{
  "contractNumber": "1/2345678/8",
  "id": "123456788",
  "person": {
    "dateOfBirth": "1995-08-21",
    "firstname": "Bart",
    "lastname": "Stark",
  },
  "premium": 99,
  "start": "2024-02-01",
}
`;
